### PR TITLE
[Packetbeat] Refactor packetbeat for use with Elastic Agent

### PR DESCRIPTION
**Pull request by andrewstucki**
## What does this PR do?

This PR refactors packetbeat to support for agent-based configuration. In order to facilitate control by agent, it does a few things:

1. Gets rid of any global variables that are initialized based on the configuration state
2. Adds handling that overrides any configuration if `packetbeat.inputs` are provided.
3. Makes the index that packetbeat pushes events to overrideable on each input (flows or protocols)
4. Uses centralized configuration management reloading to restart the sniffer and all worker goroutines when configuration changes

## Follow ups

We'll likely want to do a few things in the future, I can create issues if we want:

1. Right now, due to the way a single sniffer gets initialized and then delegates all of the protocol parsing to internally registered protocol handlers, each configuration reload, if even a single input changes, the sniffer needs to be stopped and started--so effectively all inputs are tied together. Not sure if it would make sense to move to some sort of worker routine per protocol or not, but it would get rid of some of the custom reload logic I had to add.
2. Once again, due to the single sniffer, the sniffing interface can really only be configured once. Right now I just choose the "default" interface based off of the OS, but ideally this would be customizable. That said, what it might mean is that we need to make all packetbeat integrations a single package or change packetbeat so we run a sniffer per input.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Related issues

- Related: elastic/beats#21979
- Related: elastic/beats#22145

### CODE REVIEW

File: packetbeat/beater/packetbeat.go [hunk 1]
```diff
@@ -79,8 +57,8 @@ func init() {
 	}
 }
 
-func New(b *beat.Beat, rawConfig *common.Config) (beat.Beater, error) {
-	config := config.Config{
+func initialConfig() config.Config {
```
Comment (ruflin) [inline packetbeat/beater/packetbeat.go:60]: Could the changes on packetbeat move to its own independent PR to decouple it from adding it to the agent?

**Review by ruflin — COMMENTED** [review]
We must decide, if packetbeat should be packaged by default or not. @ph @mostlyjason

File: packetbeat/beater/packetbeat.go (hunk 1 above)
Comment (andrewstucki) [reply packetbeat/beater/packetbeat.go:60]: @ruflin : done #22145

File: packetbeat/publish/publish.go [hunk 2]
```diff
@@ -84,6 +84,7 @@ func (p *TransactionPublisher) CreateReporter(
 
 	// load and register the module it's fields, tags and processors settings
 	meta := struct {
+		Index      string                  `config:"index"`
```
Comment (andrewkroh) [inline packetbeat/publish/publish.go:87]: The new `index` options should also go into the asciidoc and reference.yml configs.

**Comment by ruflin** [comment]
Would it be possible to support `inputs` on the top level instead of under the packetbeat prefix? Like this Agent could just forward.

In general, I think this PR is update to the team to get reviewed and merged as it does not tackle Agent yet, thanks for taking it out.

**Comment by andrewstucki** [comment]
@ruflin quick question about that--right now the agent configuration files do a lot more than just moving some keys around. That includes:

1. injecting agent metadata
2. adding stream dataset info into events
3. flattening `streams` input hashes into each individual input itself
4. and injecting per-index configuration for each input

having the `inputs` be at the top level doesn't mitigate any of the above, so agent still wouldn't be able to forward configuration from Kibana unmodified. We'd either still need to rely heavily on the current spec transpilation logic or add some processors into each integration package and update the configuration deserialization code in each beat to support deserializing (and using) some shared notion of a `Streams` configuration structure.

I stripped out any additions to the transpiler that would have been required to massage this into standard packetbeat configuration, but I would think that any of the above would likely be beyond the scope of getting packetbeat functioning with agent.

So, just want to clarify--just want me to un-nest `inputs` and keep the additional configuration we'd need in the spec file in place, correct?

**Comment by mostlyjason** [comment]
> 
> 
> We must decide, if packetbeat should be packaged by default or not. @ph @mostlyjason

Just checked our telemetry and packetbeat is used by about 7% of clusters. Would save a lot of bandwidth if it was downloaded on demand. This requires an internet connection or proxy though.

File: packetbeat/beater/processor.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	if p.flows != nil {
+		p.flows.Start()
+	}
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+
+		err := p.sniffer.Run()
+		if err != nil {
+			p.err <- fmt.Errorf("Sniffer loop failed: %v", err)
```
Comment (andrewkroh) [inline packetbeat/beater/processor.go:66]: Should this return after writing the error to the channel?

File: packetbeat/beater/processor.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+)
+
+type processor struct {
+	wg      sync.WaitGroup
+	flows   *flows.Flows
+	sniffer *sniffer.Sniffer
+	err     chan error
+}
+
+func newProcessor(flows *flows.Flows, sniffer *sniffer.Sniffer, err chan error) *processor {
```
Comment (andrewkroh) [inline packetbeat/beater/processor.go:44]: Who's responsible for closing the err channel? Normally the writer closes it when it's done.

File: packetbeat/beater/reloader.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	if r.runner != nil {
+		go r.runner.Stop()
+	}
+	// reinitialize runner
+	r.runner = nil
+
+	c, err := common.NewConfigFrom(config.Config)
+	if err != nil {
+		r.logger.Errorf("Unable to create new configuration for factory: %s", err)
+		return errors.Wrap(err, "Unable to create new configuration for factory")
```
Comment (andrewkroh) [inline packetbeat/beater/reloader.go:86]: Can you make the `error` messages begin with lowercase (https://github.com/golang/go/wiki/CodeReviewComments#error-strings).

File: packetbeat/config/agent.go [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+// Normalize allows the packetbeat configuration to understand
+// agent semantics
+func (c Config) Normalize() (Config, error) {
+	logp.Debug("agent", "Normalizing agent configuration")
+	if len(c.Inputs) > 0 {
+		// override everything, we're managed by agent
+		c.Flows = nil
+		c.Protocols = nil
+		c.ProtocolsList = []*common.Config{}
+		// TODO: make this configurable rather than just using the default device in
```
Comment (andrewkroh) [inline packetbeat/config/agent.go:50]: That seems important 😄 .

File: packetbeat/config/agent.go [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+// Normalize allows the packetbeat configuration to understand
+// agent semantics
+func (c Config) Normalize() (Config, error) {
+	logp.Debug("agent", "Normalizing agent configuration")
+	if len(c.Inputs) > 0 {
+		// override everything, we're managed by agent
+		c.Flows = nil
+		c.Protocols = nil
+		c.ProtocolsList = []*common.Config{}
```
Comment (andrewkroh) [inline packetbeat/config/agent.go:49]: Could `nil` be used here instead? Not sure if the old code had specials semantics for nil vs empty.

File: packetbeat/config/agent.go [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+		// TODO: make this configurable rather than just using the default device in
+		// managed mode
+		c.Interfaces.Device = defaultDevice()
+	}
+
+	for _, input := range c.Inputs {
+		if rawInputType, ok := input["type"]; ok {
+			inputType, ok := rawInputType.(string)
```
Comment (andrewkroh) [inline packetbeat/config/agent.go:57]: It looks like it could be helpful to fail with an error in case this is not a string.

File: packetbeat/beater/reloader.go (hunk 5 above)
Comment (andrewstucki) [reply packetbeat/beater/reloader.go:86]: will do, most of these error strings are an artifact of previous code, but I can do a pass an normalize anything I touched

File: packetbeat/config/agent.go (hunk 6 above)
Comment (andrewstucki) [reply packetbeat/config/agent.go:50]: Yep, definitely important. Since this code is only hit when agent-based config passes in configuration under the (undocumented) `inputs` array, I figured we could just do something sane-ish for now. I called it out in the PR description, but we'll want to discuss how we want to configure interfaces going forward since historically packetbeat creates a single sniffer on a given interface and then passes packets on to all the registered protocol dissectors after that. We'll need to decide how we can either change that or if we need to make a single protocol sniffing package that has this as a global conffiguration option sometime in the future.

File: packetbeat/config/agent.go (hunk 8 above)
Comment (andrewstucki) [reply packetbeat/config/agent.go:57]: Sure thing, I'm not too concerned about edge cases here since all of these should be coming from agent. But it would probably help surface issues quicker.

File: packetbeat/beater/processor.go (hunk 4 above)
Comment (andrewkroh) [reply packetbeat/beater/processor.go:44]: Looks like it doesn't matter whether it gets closed in this case as the reader doesn't need it to be closed to stop reading. So nothing to worry about here.

Comment (andrewstucki) [reply packetbeat/beater/processor.go:44]: The intention was actually to use this channel as a long-lived channel in the goroutine that initializes the factory (which is also the reader goroutine) and use it to propagate errors from whatever new `processor`s might get initialized due to configuration changes.

It looks like I actually forgot to close it 😅 , but so long as no more processors can get created after the channel is closed, which there shouldn't be, then we won't have issues with writing to a closed channel.

**Comment by ruflin** [comment]
For the config and inputs: My favorite solution would be that Elastic Agent just forwards the policy to packetbeat and packetbeat adds all the processors and data needed. The processors are to add important fields but packetbeat actually exactly knows which fields have to be added without Agent sending it. It would be nice if you could check on your end, how big the effort would be to directly go this step. This is what we do with apm-server and plan to also do for filebeat / metricbeat. If we don't go this step directly, we need to migrate to it at one stage.

Perhaps you can provide some examples on what config is put into the agent and what the config looks like when sent to packetbeat as yaml here. This could simplify the discussion and making sure we talk about exactly the same thing.

**Comment by andrewstucki** [comment]
Ok, so I wound up going and swapping the configuration to what I mentioned here https://github.com/elastic/beats/issues/22227#issuecomment-718159926

The changes involve some additional normalization code to inject processors into each stream. On the plus side, the reloading code becomes much simpler since it just wraps `cfgfile.RunnerList`. For the configuration, I currently enforce a single input of type `packet` to be passed in. If `packetbeat` receives anything additional inputs, it fails re-configuration. This is to punt for the time being on:

1. figuring out the nitty-gritty details of running multiple packet sniffers on a system and
2. keep the current goroutine synchronization functional (the error channels will need some work if we move to multiple sniffers)

A demo package I've been using to test all of this stuff is here:

https://github.com/andrewstucki/integrations/commit/28455c5ae52b0efce526f41c142a032780a5be95

@ruflin this approach pretty much completely removes all transpiler rules in the packetbeat spec apart from filtering and injecting agent version information (see updated #22145). We'll still need to figure out how to inject additional configuration at the `input` level in the configuration if we want to use that to control the sniffer behavior, but I figure that can wait. Let me know what you think.

**Comment by ruflin** [comment]
@andrewstucki Read your comment above last after going through all the other comments 🤦‍♂️ We should probably have all the config discussions in a single place :-) Great to hear you solved the sniffer limitation issue.

How do we get all the conversations into a single place and can drive it there to a conclusion?

**Comment by andrewstucki** [comment]
@ruflin I'm fine with doing the rest of the config discussion here and using the added unit test file `packetbeat/config/agent_test.go` to finish driving the discussion (no matter where we discuss, any spec file changes result in changes to the normalization code here, so talking about it here keeps us from having to context-switch). That test pretty shows the agent spec file configuration format of a single input (after it's been injected with agent metadata).

The issue I created and linked to (#22227) is still valid though and can be a follow up to the initial implementation, as there's still work to be done to figure out how to configure shared, sniffer-specific options.

**Review by michalpristas — APPROVED** [review]
seems to work as expected together with https://github.com/elastic/beats/pull/22145

**Review by andrewkroh — APPROVED** [review]
LGTM

File: packetbeat/beater/processor.go [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+		p.beat.Info.Name,
+		p.beat.Publisher,
+		config.IgnoreOutgoing,
+		config.Interfaces.File == "",
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	watcher := procs.ProcessesWatcher{}
```
[the remark this answers is not in the document]
Comment (andrewstucki) [reply packetbeat/beater/processor.go:126]: So the process watcher has traditionally been a global. I more or less kept it that way--making it a singleton for the entirety of the agent-level "input" (meaning 1 sniffer + 1 process watcher). I would imagine eventually we could potentially scope this based on a protocol-by-protocol basis (moving from "input" level, to "stream" level), but we'd probably need to see if that had a performance impact.

In general I tried to keep the traditionally global configuration options scoped to a single input run (1 sniffer, etc.) to make things a bit more flexible for the future and avoid introducing three different levels of configuration (global, input-level, stream-level) when configured via agent. Instead we just have input-level and stream-level, so nothing is shared across multiple inputs.

Hope that answers the question/makes sense.
